### ISSUE TITLE
SourceFilter does not work with the full file path

### ISSUE THREAD

**Issue by Irian81** [spotbugs/spotbugs#694]
Curently "spotbugs/spotbugs/src/main/java/edu/umd/cs/findbugs/filter/SourceMatcher.java" only works against plain source file name, no path only the file name, this makes the whole point of using it way less powerful than what it should be IMHO it should also consider the full file path.

Following my use case:

I was planning to exclude a set of automatically generated files that live in my maven project \target\generated-sources, from analysis, I wanted to create a generic rule when only files present in this folder and any subfolders are excluded.

To do so i created an exclude file with the following content:

&lt;FindBugsFilter&gt;
  &lt;Match&gt;
    &lt;Source name="~.&#42;generated-sources.&#42;.java"/&gt;
  &lt;/Match&gt;
&lt;/FindBugsFilter&gt;

This does not work as the SourceMatcher class only processes the file name, not the actual path.

I've attached an example maven project that demonstrates the issue, ideally I would like to exclude generated sources but retain the same warning that exists in the manually written classes.

I know I could define a package specific for generated classes, but this forces me to change the exclusion logic every time I add another package to the auto generation so not really ideal.

[testspotbugexclude.zip](https://github.com/spotbugs/spotbugs/files/2181862/testspotbugexclude.zip)

**Comment by Irian81** [comment]
I've had a bit of spare time and managed to get something working for my specific use case, 4 files modified 2 in main spotbugs (SourceMatcher, and SourceLineAnnotation), and 2 in spotbugs-maven-plugin (SpotBugsMojo, and the pom.xml to depend on 4.0.0-SNAPSHOT).

Would you be willing to review a pull request for this??

I've gone ahead and forked the 2 repos so if you want you can look at the modifications there.

To get the sample project to work it should be sufficient to add the new <addSourceDirs>true</addSourceDirs> param and up the spotbugs-maven-plugin version to a 3.1.6-SNAPSHOT built with the "patch" in place.

**Comment by ThrawnCA** [comment]
> Would you be willing to review a pull request for this??

I think we're always happy to review :).

**Comment by Irian81** [comment]
Added PR #700 for spotbugs and PR #71 for maven.

these will most probably require checking for the SourceLineAnnotation modification sanity, and at least a new test case to be added, and potentially some documentation.

Let me know if you require something else to be added

**Comment by KengoTODA** [comment]
closed by #973

**Comment by sfc-gh-akazantsev** [comment]
Hi Guys, I'm not sure if it is regression but above example provided does not work with <addSourceDirs>true</addSourceDirs>

Is there any other way to manually add source roots?

**Comment by xenoterracide** [comment]
hmmm... this doesn't seem to be working for me, my generated files are in `build/generated/sources`. looks like addSourceDirs, is only for maven? what about gradle?

```
  <Match>
    <Source name="~.*generated.*\.java" />
  </Match>
```
```
com.github.spotbugs:spotbugs-annotations:4.1.1
com.github.spotbugs:spotbugs:4.1.1
```

**Comment by mieubrisse** [comment]
I've got the same problem using the Gradle plugin v4.6.0. Immutables generates files that trip https://github.com/spotbugs/spotbugs/issues/633 , and trying to exclude them with:
```
<FindBugsFilter>
    <!-- Anything in the build directory ought to be ignored, because it includes generated files we don't have control over -->
    <Match>
        <Source name="~.*build/generated.*java"/>
    </Match>
</FindBugsFilter>
```
doesn't match, but `~.*java` does. It seems like the full filepath isn't getting included for the Gradle plugin.

**Comment by Adrodoc** [comment]
For me this works using `gradle.plugin.com.github.spotbugs.snom:spotbugs-gradle-plugin:4.6.0` and `toolVersion = '4.1.3'`. But on Windows the path seems to use `\` instead of `/`, so this is most likely operating system dependent, which is really bad. Also it uses the full path and not the path relative to the working directory or project directory. This is really bad because you don't neccessarily know where a developer checks out the code. If someone checks out their code in a directory called `generated` than suddenly a rule like `<Source name="~.*\\generated\\.*\.java"/>` would match every Java file in the whole project.

This is the rule I am using:
```xml
<Match>
  <Source name="~.*\\src-gen\\.*"/>
</Match>
```
